## Python Final Project
### Kartik, Sahil and Manisha

In [1]:
import pandas as pd
from alpha_vantage.timeseries import TimeSeries
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import requests

**Loading COVID19 datasets**

In [9]:
confirmed_cases_url = "https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_time_series/time_series_covid19_confirmed_global.csv"

death_cases_url = "https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_time_series/time_series_covid19_deaths_global.csv"

cases = pd.read_csv(confirmed_cases_url)
deaths = pd.read_csv(death_cases_url)

In [10]:
cases

Province/State        Country/Region        Lat        Long  1/22/20  \
0              NaN           Afghanistan  33.939110   67.709953        0   
1              NaN               Albania  41.153300   20.168300        0   
2              NaN               Algeria  28.033900    1.659600        0   
3              NaN               Andorra  42.506300    1.521800        0   
4              NaN                Angola -11.202700   17.873900        0   
..             ...                   ...        ...         ...      ...   
284            NaN    West Bank and Gaza  31.952200   35.233200        0   
285            NaN  Winter Olympics 2022  39.904200  116.407400        0   
286            NaN                 Yemen  15.552727   48.516388        0   
287            NaN                Zambia -13.133897   27.849332        0   
288            NaN              Zimbabwe -19.015438   29.154857        0   

     1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  ...  2/28/23  3/1/23  \
0          0        0        0        0        0  ...   209322  209340   
1          0        0        0        0        0  ...   334391  334408   
2          0        0        0        0        0  ...   271441  271448   
3          0        0        0        0        0  ...    47866   47875   
4          0        0        0        0        0  ...   105255  105277   
..       ...      ...      ...      ...      ...  ...      ...     ...   
284        0        0        0        0        0  ...   703228  703228   
285        0        0        0        0        0  ...      535     535   
286        0        0        0        0        0  ...    11945   11945   
287        0        0        0        0        0  ...   343012  343012   
288        0        0        0        0        0  ...   263921  264127   

     3/2/23  3/3/23  3/4/23  3/5/23  3/6/23  3/7/23  3/8/23  3/9/23  
0    209358  209362  209369  209390  209406  209436  209451  209451  
1    334408  334427  334427  334427  334427  334427  334443  334457  
2    271463  271469  271469  271477  271477  271490  271494  271496  
3     47875   47875   47875   47875   47875   47875   47890   47890  
4    105277  105277  105277  105277  105277  105277  105288  105288  
..      ...     ...     ...     ...     ...     ...     ...     ...  
284  703228  703228  703228  703228  703228  703228  703228  703228  
285     535     535     535     535     535     535     535     535  
286   11945   11945   11945   11945   11945   11945   11945   11945  
287  343079  343079  343079  343135  343135  343135  343135  343135  
288  264127  264127  264127  264127  264127  264127  264276  264276  

[289 rows x 1147 columns]

In [11]:
deaths

Province/State        Country/Region        Lat        Long  1/22/20  \
0              NaN           Afghanistan  33.939110   67.709953        0   
1              NaN               Albania  41.153300   20.168300        0   
2              NaN               Algeria  28.033900    1.659600        0   
3              NaN               Andorra  42.506300    1.521800        0   
4              NaN                Angola -11.202700   17.873900        0   
..             ...                   ...        ...         ...      ...   
284            NaN    West Bank and Gaza  31.952200   35.233200        0   
285            NaN  Winter Olympics 2022  39.904200  116.407400        0   
286            NaN                 Yemen  15.552727   48.516388        0   
287            NaN                Zambia -13.133897   27.849332        0   
288            NaN              Zimbabwe -19.015438   29.154857        0   

     1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  ...  2/28/23  3/1/23  \
0          0        0        0        0        0  ...     7896    7896   
1          0        0        0        0        0  ...     3598    3598   
2          0        0        0        0        0  ...     6881    6881   
3          0        0        0        0        0  ...      165     165   
4          0        0        0        0        0  ...     1933    1933   
..       ...      ...      ...      ...      ...  ...      ...     ...   
284        0        0        0        0        0  ...     5708    5708   
285        0        0        0        0        0  ...        0       0   
286        0        0        0        0        0  ...     2159    2159   
287        0        0        0        0        0  ...     4057    4057   
288        0        0        0        0        0  ...     5663    5668   

     3/2/23  3/3/23  3/4/23  3/5/23  3/6/23  3/7/23  3/8/23  3/9/23  
0      7896    7896    7896    7896    7896    7896    7896    7896  
1      3598    3598    3598    3598    3598    3598    3598    3598  
2      6881    6881    6881    6881    6881    6881    6881    6881  
3       165     165     165     165     165     165     165     165  
4      1933    1933    1933    1933    1933    1933    1933    1933  
..      ...     ...     ...     ...     ...     ...     ...     ...  
284    5708    5708    5708    5708    5708    5708    5708    5708  
285       0       0       0       0       0       0       0       0  
286    2159    2159    2159    2159    2159    2159    2159    2159  
287    4057    4057    4057    4057    4057    4057    4057    4057  
288    5668    5668    5668    5668    5668    5668    5671    5671  

[289 rows x 1147 columns]

In [6]:
deaths['Province/State'].isna().sum(), cases['Province/State'].isna().sum()

(198, 198)

The ``Province/State`` column does not look very good, with 198 ``NaN`` values out of 289, in both of the dataframes. However, we can not remove those rows altogether as they might be very useful for more specific further analysis.
</b>
Let us now have a bigger picture of the dataset by aggregating deaths and cases over all countries.

In [7]:
global_summed = pd.DataFrame([cases.iloc[:,4: ].sum(axis=0), deaths.iloc[:,4: ].sum(axis=0)], index=['cases', 'deaths'])
global_summed

1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  1/28/20  \
cases       557      657      944     1437     2120     2929     5580   
deaths       17       18       26       42       56       82      131   

        1/29/20  1/30/20  1/31/20  ...    2/28/23     3/1/23     3/2/23  \
cases      6169     8237     9927  ...  675322238  675542852  675731911   
deaths      133      172      214  ...    6872682    6874463    6876031   

           3/3/23     3/4/23     3/5/23     3/6/23     3/7/23     3/8/23  \
cases   675914580  675968775  676024901  676082941  676213378  676392824   
deaths    6877325    6877601    6877749    6878115    6879038    6880483   

           3/9/23  
cases   676570149  
deaths    6881802  

[2 rows x 1143 columns]

The dataset is in what is called the "wide" form, let us tidy it up a little for better analysis.

In [8]:
covid = global_summed.transpose()
covid.index = pd.to_datetime(covid.index)
covid

cases   deaths
2020-01-22        557       17
2020-01-23        657       18
2020-01-24        944       26
2020-01-25       1437       42
2020-01-26       2120       56
...               ...      ...
2023-03-05  676024901  6877749
2023-03-06  676082941  6878115
2023-03-07  676213378  6879038
2023-03-08  676392824  6880483
2023-03-09  676570149  6881802

[1143 rows x 2 columns]

The dataset looks pretty apt, although minor cleaning might be required which we will carry out once the complete data has been extracted.

**Loading Stocks datasets**

Further, we will extract stock market prices for each of these dates, so as to analyse COVID19's effect on these prices. We use the following 5 stock indices for the categories listed in question.
1. Overall American Market: **NASDAQ**
2. Overall Canadian Market: **Toronto Stock Exchange**
3. Travel Sector: **AirBNB** (AlphaVantage did not seem to have an overall index for this sector, so we are using AirBNB for reference)
4. Real Estate Sector: **MSCI U.S. REIT Index**
5. Precious Metals: **S&P GSCI**

In [63]:
API_KEY = 'FH3KFFPOMYNI13R9'
START_DATE = covid.index[0].strftime('%Y-%m-%d')
END_DATE = covid.index[-1].strftime('%Y-%m-%d')
url = 'https://www.alphavantage.co/query?function=TIME_SERIES_DAILY_ADJUSTED&symbol={sym}&outputsize=full&apikey={key}&datatype=csv'

In [78]:

def conv_hi_lo(df):
    df.index = pd.to_datetime(df['timestamp'])
    return df[(df.index>START_DATE) & (df.index<END_DATE)].loc[:, ['high', 'low']]

In [79]:
nasdaq_data = conv_hi_lo(pd.read_csv(url.format(key=API_KEY, sym='NDAQ')))
tsx_data = conv_hi_lo(pd.read_csv(url.format(key=API_KEY, sym='TSE')))
realest_data = conv_hi_lo(pd.read_csv(url.format(key=API_KEY, sym='MSCI')))
metals_data = conv_hi_lo(pd.read_csv(url.format(key=API_KEY, sym='S&PGSCI')))
travel_data = conv_hi_lo(pd.read_csv(url.format(key=API_KEY, sym='ABNB')))

In [80]:
nasdaq_data

high     low
timestamp                  
2023-03-08   55.395   54.82
2023-03-07   56.215   54.95
2023-03-06   56.710   55.96
2023-03-03   56.325   55.80
2023-03-02   56.110   55.06
...             ...     ...
2020-01-29  115.770  109.76
2020-01-28  111.590  109.80
2020-01-27  110.280  109.00
2020-01-24  110.290  109.24
2020-01-23  109.680  108.48

[787 rows x 2 columns]

Finally, we have all the stocks datasets with desired columns of daily low and daily high within the date ranges which are in the ``covid`` dataframe.
However, there appears to be a minor problem with the ``covid`` dataframe itself, the ``deaths`` and ``cases`` are commulative numbers. We want them to be daily numbers.
Let's create new columns for non-commulative stats.

In [87]:
covid.rename({'deaths':'commulative_deaths', 'cases':'commulative_cases'}, axis=1, inplace=True)
covid['cases'], covid['deaths'] = covid['commulative_cases'].diff(), covid['commulative_deaths'].diff()
covid = covid.fillna(0)
covid

commulative_cases  commulative_deaths     cases  deaths
2020-01-22                557                  17       0.0     0.0
2020-01-23                657                  18     100.0     1.0
2020-01-24                944                  26     287.0     8.0
2020-01-25               1437                  42     493.0    16.0
2020-01-26               2120                  56     683.0    14.0
...                       ...                 ...       ...     ...
2023-03-05          676024901             6877749   56126.0   148.0
2023-03-06          676082941             6878115   58040.0   366.0
2023-03-07          676213378             6879038  130437.0   923.0
2023-03-08          676392824             6880483  179446.0  1445.0
2023-03-09          676570149             6881802  177325.0  1319.0

[1143 rows x 4 columns]

There appears to be another problem before merging the datasets, there are some dates missing from the datasets extracted from AlphaVantage. The sizes of ``covid`` and stock datasets do not match.
We are going to ``reindex()`` the datasets from ``START_DATE`` to ``END_DATE`` and interpolate the missing dates with stocks of previous date

In [100]:
len(nasdaq_data)==len(covid)

False